### ISSUE TITLE
Time to first spike

### ISSUE THREAD

**Issue by Dola47** [norse/notebooks#9]
First of all, thank you for your very informative tutorials.

Then I would like to kick off by asking what is the difference between the **SpikeLatencyLIFEncoder** and **ConstantCurrentLIFEncoder** because I see that both can be considered as a type of Temporal encoding, but I do not get what is the main difference between both of them.

A couple of questions here:

> There are other options of course, for example, to consider the average membrane voltage in a given time window or **use a LIF neuron output layer and consider the time to first spike**.

-  Can you provide an example of how the time of the first spike can be considered?
- Moreover, when you used the LatencyEncoder in the MNIST tutorial, I see that the network kept learning rate encoded output. You have not changed anything it is the same architecture, the same decoder, just changed the encoding. However, 
I think the network is still expecting firing at each time step, does that make sense?
- What I understand that when we use Latency encoding then we should not consider all of the time steps but consider only the time steps where the correct class spikes, so our loss function should encourage the correct class to fire first and the incorrect classes to fire later. And of course, then our accuracy function should be changed to count only the classes the fire first. I do not have any clue how that can be applied for example to the MNIST dataset. However, it will be really much appreciated if you can show me how this can be done.

Thanks and looking forward to your response.

**Comment by Jegp** [comment]
Thanks for the question. I made this quick visualization to illustrate the difference with a digit from the MNIST dataset. The left representation is dense, the next is a spike latency encoding, and the final is a constant current LIF encoding. 

![grafik](https://user-images.githubusercontent.com/1064317/115868561-ef04e300-a43c-11eb-988c-6d1b82e0fc3e.png)

The spike latency encodes the *first time to spike* for each neuron. Meaning, there will only be one spike per neuron. But in the constant current case, the LIF neurons is injected with a constant current, which means that they'll keep spiking with a frequency  based on the injected current.

>     * Moreover, when you used the LatencyEncoder in the MNIST tutorial, I see that the network kept learning rate encoded output. You have not changed anything it is the same architecture, the same decoder, just changed the encoding. However,
>       I think the network is still expecting firing at each time step, does that make sense?

I'm not sure what you mean that the network is learning a rate based output. If you use latency encoding, the network is only receiving the *first* spike from each input neuron, and nothing else (for that single data point). That is indeed different from constantly injecting spikes with a fixed rate, but the network should be able to learn from the single spike signal (almost) just as well. BPTT should take care of exactly how the weights will have to be updated to accommodate for this.

Was that helpful?

**Comment by Dola47** [comment]
Hi @Jegp,

Thanks a lot for your reply.

First of all, can you provide some illustration of how BPTT is being done here?  For example in the mnist_classifiers tutorial, your decoder only considers the maximum voltages to calculate the softmax. 
```
def decode(x):
    x, _ = torch.max(x, 0)
    log_p_y = torch.nn.functional.log_softmax(x, dim=1)
    return log_p_y
```
Then I think we are not considering all of the time steps when you calculate the nll_loss, am I right? And how then the BPTT is being calculated?

For me, for example, if our voltages are of the shape [48, 256, 10]. I would expect to have something like that to calculate the loss 
```
for step in range(49):
  loss_val += loss_fn(log_p_y[step], targets_it)
```
Where `loss_fn` is the `NLLLoss()`, and `log_p_y` is `log_softmax_fn`.  And by that, we are summing the loss over time to perform BPTT. And that's where I am also saying our network is following a **rate encoding** approach because an output target is expected at every step (one output class must first at every step). And that will not be appropriate with Time to first spike encoding. That's why I raised the first question as well. 

Finally, for the accuracy, I think if we have done rate encoding then I think we should count the total number of output spikes as a measure of accuracy, however, you are not considering the number of spikes at all in your implementation. On the other hand, for the temporal encoding, we should only count the class that fires first.

I might have a misunderstanding so please correct me.

To sum:

- Inform me how BPTT is being done in your implementation?
- How you consider rate encoding, and how you consider its accuracy?
- How you consider temporal encoding, how it should be decoded in the most appropriate way, and how you consider its accuracy?

I hope I am clear enough thanks a lot and looking forward to your response.

**Comment by Jegp** [comment]
> * Inform me how BPTT is being done in your implementation?

BPTT is being done by PyTorch. PyTorch is basically building up a large graph of the gradients given the computational graph of your model. This tutorial is explaining it very well: https://pytorch.org/tutorials/beginner/basics/autogradqs_tutorial.html

> * How you consider rate encoding, and how you consider its accuracy?

Rate encoding is one way to convert a pixel image into a temporal tensor. The Poisson encoder is one example of this. If you also decode with rates, then you will of course "train" the network to work with rates. The `max` operation only looks at the maximum voltage for all the timesteps. I wouldn't call that rate-based, because it doesn't care about the frequency.

> * How you consider temporal encoding, how it should be decoded in the most appropriate way, and how you consider its accuracy?

Good question! Ideally, you wouldn't have any static input (like MNIST). Instead, you would have a constant flow of events like in [event-based vision](https://github.com/uzh-rpg/event-based_vision_resources). We don't always have access to that kind of data of course, so you could also apply spike latency encoding. I would argue that's a temporal coding because there's no periodicity in the encoded output.

It's great you have so many questions! I would recommend using the [GitHub discussions](https://github.com/norse/norse/discussions) or our [Discord server](https://discord.gg/7fGN359) for questions that aren't exactly _issues_ with the code/notebooks.

I'll close the issue for now. Feel free to reopen if you come across unexpected behaviour.

**Comment by Dola47** [comment]
> Thanks for the question. I made this quick visualization to illustrate the difference with a digit from the MNIST dataset. The left representation is dense, the next is a spike latency encoding, and the final is a constant current LIF encoding.
> 
> ![grafik](https://user-images.githubusercontent.com/1064317/115868561-ef04e300-a43c-11eb-988c-6d1b82e0fc3e.png)
> 
> The spike latency encodes the _first time to spike_ for each neuron. Meaning, there will only be one spike per neuron. But in the constant current case, the LIF neurons is injected with a constant current, which means that they'll keep spiking with a frequency based on the injected current.
> 
> > ```
> > * Moreover, when you used the LatencyEncoder in the MNIST tutorial, I see that the network kept learning rate encoded output. You have not changed anything it is the same architecture, the same decoder, just changed the encoding. However,
> >   I think the network is still expecting firing at each time step, does that make sense?
> > ```
> 
> I'm not sure what you mean that the network is learning a rate based output. If you use latency encoding, the network is only receiving the _first_ spike from each input neuron, and nothing else (for that single data point). That is indeed different from constantly injecting spikes with a fixed rate, but the network should be able to learn from the single spike signal (almost) just as well. BPTT should take care of exactly how the weights will have to be updated to accommodate for this.
> 
> Was that helpful?

Hi @Jegp,

Can you provide the source code for producing the above plots?

Thanks.
